xdsl/transforms/loop_invariant_code_motion.py

```diff
@@ -0,0 +1,43 @@
+"""
+This pass hoists operation that are invariant to the loops.
+
+Similar to MLIR's loop invariant code motion:
+
+https://mlir.llvm.org/doxygen/LoopInvariantCodeMotion_8cpp_source.html
+
+An operation is loop-invariant if it depends only of values defined outside of the loop.
+LICM moves these operations out of the loop body so that they are not computed more than
+once.
+
+  for i in range(x, N, M):                for i in range(x, N, M):
+    for j in range(0, M, K):    ---->        c[i]= A[1] + b[1]
+      c[i]=A[1]+b[1]
+"""
+
+from xdsl.ir import Operation, Region
+from xdsl.traits import IsTerminator
+
+
+def can_be_hoisted(op: Operation, target_region: Region) -> bool | None:
+    """
+    Checks whether the given op can be hoisted by checking that
+    - the op and none of its contained operations depend on values inside of the
+     loop (by means of calling definedOutside).
+    - the op has no side-effects.
```

Comment:
This seems to not be tested in this function, let's remove it?

Reply:
Done